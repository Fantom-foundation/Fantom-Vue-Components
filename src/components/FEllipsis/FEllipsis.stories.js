import FEllipsis from './FEllipsis.vue';

export default {
    title: 'FEllipsis',
    component: FEllipsis,
};

const resizableBox = 'max-width: 900px; resize: horizontal; overflow: auto; padding: 24px; border: 1px solid #eee;';

export const Default = () => ({
    components: { FEllipsis },
    template: `
        <div :style="resizableBox">
            <f-ellipsis text="0x2fF3c477b65B37678B8DB93A36426Aa2b6B1AcF9A1f12B766C023Ab1807F44cDC5B1e5746dC69d7e" />
        </div>
    `,
    data() {
        return {
            resizableBox,
        };
    },
});

export const OverflowStart = () => ({
    components: { FEllipsis },
    template: `
        <div :style="resizableBox">
            <f-ellipsis
                text="0x2fF3c477b65B37678B8DB93A36426Aa2b6B1AcF9A1f12B766C023Ab1807F44cDC5B1e5746dC69d7e"
                overflow="start"
            />
        </div>
    `,
    data() {
        return {
            resizableBox,
        };
    },
});

export const OverflowMiddleEnd = () => ({
    components: { FEllipsis },
    template: `
        <div :style="resizableBox">
            <f-ellipsis
                text="0x2fF3c477b65B37678B8DB93A36426Aa2b6B1AcF9A1f12B766C023Ab1807F44cDC5B1e5746dC69d7e"
                overflow="middle"
                :fixed-chars-count="7"
            />
        </div>
    `,
    data() {
        return {
            resizableBox,
        };
    },
});

export const OverflowMiddleStart = () => ({
    components: { FEllipsis },
    template: `
        <div :style="resizableBox">
            <f-ellipsis
                text="0x2fF3c477b65B37678B8DB93A36426Aa2b6B1AcF9A1f12B766C023Ab1807F44cDC5B1e5746dC69d7e"
                overflow="middle"
                :fixed-chars-count="7"
                fixed-chars-pos="start"
            />
        </div>
    `,
    data() {
        return {
            resizableBox,
        };
    },
});

export const Slots = () => ({
    components: { FEllipsis },
    template: `
        <div :style="resizableBox">
            <f-ellipsis
                text="0x2fF3c477b65B37678B8DB93A36426Aa2b6B1AcF9A1f12B766C023Ab1807F44cDC5B1e5746dC69d7e"
                overflow="middle"
                :fixed-chars-count="7"
                fixed-chars-pos="start"
            >
                <template #prefix><b>prefix</b></template>
                <template #suffix><b>suffix</b></template>
            </f-ellipsis>
        </div>
    `,
    data() {
        return {
            resizableBox,
        };
    },
});
