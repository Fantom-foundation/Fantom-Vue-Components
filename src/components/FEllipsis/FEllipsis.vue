<template>
    <span class="f-ellipsis" :class="cssClass">
        <span><slot name="prefix"></slot></span>

        <template v-if="overflow === 'middle'">
            <span class="left-chars">{{ leftChars }} </span>
            <span class="right-chars">{{ rightChars }} </span>
        </template>
        <template v-else>
            {{ cText }}
        </template>

        <span><slot name="suffix"></slot></span>
    </span>
</template>

<script>
/**
 * Renders ellipsis (using css text-overflow property) at the start, at the middle or at the end of the given text.
 */
export default {
    name: 'FEllipsis',

    props: {
        text: {
            type: String,
            default: '',
            required: true,
        },
        /**
         * Position of three dots. (`'start'` | `'middle'` | `'end'`)
         *
         * @type {('start' | 'middle' | 'end')}
         */
        overflow: {
            type: String,
            default: 'end',
            validator: function(_value) {
                return ['start', 'middle', 'end'].indexOf(_value) !== -1;
            },
        },
        /** Number of fixed characters (at the start or at the end), if `overflow` is set to `'middle'`. */
        fixedCharsCount: {
            type: Number,
            default: 5,
            validator(_value) {
                return _value > 0;
            },
        },
        /**
         * Position of fixed characters, if `overflow` is set to `'middle'`.
         *
         * @type {('start' | 'end')}
         */
        fixedCharsPos: {
            type: String,
            default: 'end',
            validator(_value) {
                return ['start', 'end'].indexOf(_value) !== -1;
            },
        },
        /**
         * Align text to the right.
         */
        alignRight: {
            type: Boolean,
            default: false,
        },
    },

    computed: {
        cText() {
            return this.text;
        },
        /**
         * Get left part of text.
         */
        leftChars() {
            const { cText } = this;

            if (this.fixedCharsPos === 'end') {
                return cText.substring(0, cText.length - this.fixedCharsCount);
            }

            return cText.substring(0, this.fixedCharsCount);
        },
        /**
         * Get right part of text.
         */
        rightChars() {
            const { cText } = this;

            if (this.fixedCharsPos === 'end') {
                return cText.substring(cText.length - this.fixedCharsCount);
            }

            return cText.substring(this.fixedCharsCount);
        },

        cssClass() {
            return [
                `overflow-${this.overflow}`,
                {
                    'fixed-char-pos-start': this.fixedCharsPos === 'start',
                    'f-ellipsis-align-right': this.alignRight,
                },
            ];
        },
    },
};
</script>

<style lang="scss">
@import 'style';
</style>
