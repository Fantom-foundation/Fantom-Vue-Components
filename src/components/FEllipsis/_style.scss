.f-ellipsis {
    display: block;

    &:not(.overflow-middle) {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &.overflow-start {
        direction: rtl;
        text-align: end;
    }

    &.overflow-middle {
        display: flex;

        .right-chars {
            white-space: nowrap;
        }

        &:not(.fixed-char-pos-start) > .left-chars {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
        }

        &.fixed-char-pos-start > .right-chars {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            direction: rtl;
            text-align: end;
        }

        &.f-ellipsis-align-right {
            justify-content: flex-end;
        }
    }
}

html[dir="rtl"], body[dir="rtl"] {
    .f-ellipsis {
        &.overflow-start {
            direction: ltr;
        }

        &.fixed-char-pos-start > .right-chars {
            direction: ltr;
        }
    }
}
