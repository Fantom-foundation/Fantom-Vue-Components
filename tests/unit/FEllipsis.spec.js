import { mount } from '@vue/test-utils';
import FEllipsis from '@/components/FEllipsis/FEllipsis.vue';

let wrapper = null;

function destroyWrapper() {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper() {
    wrapper = mount(FEllipsis, {
        propsData: {
            text: '123456789',
            overflow: 'middle',
            fixedCharsPos: 'start',
        },
    });
}

beforeEach(() => {
    createWrapper();
});

afterEach(() => {
    destroyWrapper();
});

describe('FEllipsis', () => {
    it('should render correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should return correct left part of the text', () => {
        expect(wrapper.vm.leftChars).toBe('12345');
    });

    it('should return correct right part of the text', () => {
        expect(wrapper.vm.rightChars).toBe('6789');
    });
});
