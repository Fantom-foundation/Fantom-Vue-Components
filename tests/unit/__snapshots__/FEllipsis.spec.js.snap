// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FEllipsis should render correctly 1`] = `
<span
  class="f-ellipsis overflow-middle fixed-char-pos-start"
>
  <span />
   
  <span
    class="left-chars"
  >
    12345 
  </span>
   
  <span
    class="right-chars"
  >
    6789 
  </span>
   
  <span />
</span>
`;
